src/cgame/cg_rocket_draw.cpp

```diff
@@ -2174,6 +2174,130 @@ class PredictedMineEfficiencyElement : public HudElement
 	std::unordered_map<int, std::string> pluralSuffix;
 };
 
+class BarbsHudElement : public HudElement
+{
+public:
+	BarbsHudElement ( const Rocket::Core::String& tag ) :
+	HudElement ( tag, ELEMENT_ALIENS ),
+	numBarbs( 0 ),
+	maxBarbs( BG_Weapon( WP_ALEVEL3_UPG )->maxAmmo ),
+	regenerationInterval ( 0 ),
+	t0 ( 0 ),
+	offset ( 0 ) {}
+
+	void OnAttributeChange( const Rocket::Core::AttributeNameList& changed_attributes )
+	{
+		HudElement::OnAttributeChange( changed_attributes );
+		if ( changed_attributes.find( "src" ) != changed_attributes.end() )
+		{
+			if ( maxBarbs > 0 )
+			{
+				Rocket::Core::String src = GetAttribute<Rocket::Core::String>( "src", "" );
+				Rocket::Core::String base( va("<img class='barbs' src='%s' />", src.CString() ) );
+				Rocket::Core::String rml;
+
+				for ( int i = 0; i < maxBarbs; i++ )
+				{
+					rml += base;
+				}
+				SetInnerRML( rml );
+			}
+			else
+			{
+				SetInnerRML( "" );
+			}
+		}
+	}
+
+	void DoOnUpdate()
+	{
+		int newNumBarbs = cg.snap->ps.ammo;
+		int interval = BG_GetBarbRegenerationInterval( cg.snap->ps );
+
+		if ( newNumBarbs < maxBarbs )
+		{
+			// start regenerating barb now
+			if ( newNumBarbs > numBarbs || ( newNumBarbs < numBarbs && numBarbs == maxBarbs ) )
+			{
+				t0 = cg.time;
+				// sin(-pi/2) is minimal
+				offset = -M_PI_2;
+				assert( GetSin() == -1.0 );
```

Comment:
Personally I wouldn't spam asserts for stuff that's mainly visual and not really critical. In particular I wouldn't expect exact floats on all platforms.